controllers/datadogagent/controller_reconcile_agent.go

```diff
@@ -238,3 +240,78 @@ func (r *Reconciler) generateNodeAffinity(p string, affinity *corev1.Affinity) *
 
 	return affinity
 }
+
+func (r *Reconciler) handleProviders(ctx context.Context) error {
+	if err := r.updateProviderStore(ctx); err != nil {
+		return err
+	}
+
+	if err := r.cleanupUnusedProviders(ctx); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// updateProviderStore recomputes the required providers by making a
+// call to the apiserver for an updated list of nodes
+func (r *Reconciler) updateProviderStore(ctx context.Context) error {
+	nodeList := corev1.NodeList{}
+	err := r.client.List(ctx, &nodeList)
+	if err != nil {
+		return err
+	}
+
+	// recompute providers using updated node list
+	providersList := make(map[string]struct{})
+	for _, node := range nodeList.Items {
+		provider := kubernetes.DetermineProvider(node.Labels)
+		if _, ok := providersList[provider]; !ok {
+			providersList[provider] = struct{}{}
+		}
+	}
+
+	r.providers.SetAllProviders(providersList)
```

Comment:
for better intuition could we use the reconciler parameter `providerStore` instead of `providers` ? (i can add this comment to the other PR)